In [ ]:
import pandas as pd

df = pd.read_csv('survey_results_public.csv', low_memory=False, encoding='utf-8', on_bad_lines='skip')
schema = pd.read_csv('survey_results_schema.csv', encoding='utf-8')

df.head()

,ResponseId,MainBranch,Age,EdLevel,Employment,EmploymentAddl,WorkExp,LearnCodeChoose,LearnCode,LearnCodeAI,...,AIAgentOrchestration,AIAgentOrchWrite,AIAgentObserveSecure,AIAgentObsWrite,AIAgentExternal,AIAgentExtWrite,AIHuman,AIOpen,ConvertedCompYearly,JobSat
0,1,I am a developer by profession,25-34 years old,"Master’s degree (M.A., M.S., M.Eng., MBA, etc.)",Employed,"Caring for dependents (children, elderly, etc.)",8.0,"Yes, I am not new to coding but am learning ne...",Online Courses or Certification (includes all ...,"Yes, I learned how to use AI-enabled tools for...",...,Vertex AI,NaN,NaN,NaN,ChatGPT,NaN,When I don’t trust AI’s answers,"Troubleshooting, profiling, debugging",61256.0,10.0
1,2,I am a developer by profession,25-34 years old,"Associate degree (A.A., A.S., etc.)",Employed,NaN,2.0,"Yes, I am not new to coding but am learning ne...",Online Courses or Certification (includes all ...,"Yes, I learned how to use AI-enabled tools for...",...,NaN,NaN,NaN,NaN,NaN,NaN,When I don’t trust AI’s answers;When I want to...,All skills. AI is a flop.,104413.0,9.0
2,3,I am a developer by profession,35-44 years old,"Bachelor’s degree (B.A., B.S., B.Eng., etc.)","Independent contractor, freelancer, or self-em...",None of the above,10.0,"Yes, I am not new to coding but am learning ne...",Online Courses or Certification (includes all ...,"Yes, I learned how to use AI-enabled tools for...",...,NaN,NaN,NaN,NaN,ChatGPT;Claude Code;GitHub Copilot;Google Gemini,NaN,When I don’t trust AI’s answers;When I want to...,"Understand how things actually work, problem s...",53061.0,8.0
3,4,I am a developer by profession,35-44 years old,"Bachelor’s degree (B.A., B.S., B.Eng., etc.)",Employed,None of the above,4.0,"Yes, I am not new to coding but am learning ne...","Other online resources (e.g. standard search, ...","Yes, I learned how to use AI-enabled tools for...",...,NaN,NaN,NaN,NaN,ChatGPT;Claude Code,NaN,When I don’t trust AI’s answers;When I want to...,NaN,36197.0,6.0
4,5,I am a developer by profession,35-44 years old,"Master’s degree (M.A., M.S., M.Eng., MBA, etc.)","Independent contractor, freelancer, or self-em...","Caring for dependents (children, elderly, etc.)",21.0,"No, I am not new to coding and did not learn n...",NaN,"Yes, I learned how to use AI-enabled tools for...",...,NaN,NaN,NaN,NaN,NaN,NaN,When I don’t trust AI’s answers,"critical thinking, the skill to define the tas...",60000.0,7.0


In [ ]:
# Задание 1 — количество респондентов
total_respondents = len(df)

print("Total respondents:", total_respondents)

Total respondents: 49191


**Вывод:** В опросе приняли участие 40191 респондент. Это достаточно большой объём данных для проведения анализа.

In [ ]:
# Задание 2 — полный ответы

questions = set(schema['qname'])
columns = set(df.columns)

valid_columns = list(questions.intersection(columns))

full_answers = df[valid_columns].dropna(how='any')

print("Respondents with full answers:", len(full_answers))

Respondents with full answers: 0


**Вывод:** В датасете нет респондентов, которые ответили на все вопросы. Это связано с большим количеством вопросов и пропущенных значений.

In [ ]:
# Задание 3 — статистика WorkExp

work_exp = df['WorkExp'].dropna()

print("Mean:", work_exp.mean())
print("Median:", work_exp.median())
print("Mode:", work_exp.mode()[0])

Mean: 13.367402606485907
Median: 10.0
Mode: 10.0


**Вывод:** Средний опыт работы составляет около 13 лет, медианное и наиболее частое значение — 10 лет. Это говорит о том, что большинство респондентов имеют около 10 лет опыта.

In [ ]:
# Задание 4 — удалённая работа

df['RemoteWork'].value_counts()

remote_workers = df[df['RemoteWork'] == 'Remote'].copy()

print("Remote workers:", len(remote_workers))

Remote workers: 10931


**Вывод:** В датасете 10931 респондент работает полностью удалённо. Это показывает значительную долю специалистов, выбирающих remote формат работы.

In [ ]:
# Задание 5 — Python %

python_users = df['LanguageHaveWorkedWith'].str.contains('Python', na=False)

python_percent = round(python_users.mean() * 100, 2)

print("Python users:", python_percent, "%")

Python users: 37.54 %


**Вывод:** Около 37.54% респондентов используют Python. Это подтверждает высокую популярность языка среди разработчиков.

In [ ]:
# Задание 6 — онлайн обучение

online_users = df['LearnCode'].str.contains('Online', case=False, na=False)

print("Users who learned online:", online_users.sum())

Users who learned online: 28352


**Вывод:** 28352 респондента обучались программированию с помощью онлайн-курсов.
Это показывает, что онлайн-обучение является одним из самых популярных способов изучения программирования.

In [ ]:
# Задание 7 — Географический анализ зарплат Python-разработчиков

# фильтруем Python-разработчиков
python_dev = df[df['LanguageHaveWorkedWith'].str.contains('Python', na=False)]

# группируем по странам и считаем среднее и медиану
salary_stats = python_dev.groupby('Country')['ConvertedCompYearly'].agg(['mean', 'median'])

# убираем пропуски
salary_stats = salary_stats.dropna()

# округляем и сортируем
salary_stats = salary_stats.round(2).sort_values(by='mean', ascending=False)

salary_stats.head()

,mean,median
Country,,
Oman,390135.00,390135.0
Andorra,226103.50,226103.5
Viet Nam,218837.17,8254.0
United States of America,173298.59,150000.0
Switzerland,156456.60,142592.0


**Вывод:** Уровень заработной платы Python-разработчиков значительно отличается в зависимости от страны. В некоторых странах наблюдается большой разрыв между средним и медианным значением, что может указывать на неравномерное распределение доходов.

In [ ]:
# Задание 8 — топ 5 по зарплате

salary_df = df[['ConvertedCompYearly', 'EdLevel']].dropna(subset=['ConvertedCompYearly'])

top5_education = salary_df.sort_values(by='ConvertedCompYearly', ascending=False).head(5)

top5_education

,ConvertedCompYearly,EdLevel
34267,50000000.0,"Associate degree (A.A., A.S., etc.)"
28700,33552715.0,"Master’s degree (M.A., M.S., M.Eng., MBA, etc.)"
43143,18387548.0,"Associate degree (A.A., A.S., etc.)"
35353,15430267.0,"Bachelor’s degree (B.A., B.S., B.Eng., etc.)"
45971,13921760.0,"Master’s degree (M.A., M.S., M.Eng., MBA, etc.)"


**Вывод:** Среди самых высокооплачиваемых респондентов встречаются как бакалавры,
так и специалисты с магистерским образованием. Это показывает, что высокий доход возможен при разных уровнях образования.

In [ ]:
# Задание 9 — Анализ популярности Python по возрастным группам

# создаём копию DataFrame (чтобы избежать SettingWithCopyWarning)
df_copy = df[['Age', 'LanguageHaveWorkedWith']].copy()

# создаём столбец с признаком использования Python
df_copy['UsesPython'] = df_copy['LanguageHaveWorkedWith'].str.contains('Python', na=False)

# группируем по возрасту и считаем процент
python_by_age = df_copy.groupby('Age')['UsesPython'].mean() * 100

# округляем значения
python_by_age = python_by_age.round(2)

python_by_age

,UsesPython
Age,
18-24 years old,40.00
25-34 years old,36.94
35-44 years old,36.72
45-54 years old,38.63
55-64 years old,37.24
65 years or older,31.63
Prefer not to say,31.22


**Вывод:** Python наиболее популярен среди молодых специалистов (18–24 года), где его используют около 40% респондентов. С увеличением возраста популярность Python постепенно снижается.

In [ ]:
# Задание 10 — индустрии топ-remote

# считаем 75 перцентиль
threshold = df['ConvertedCompYearly'].quantile(0.75)

# фильтруем: зарплата выше порога + удалёнка
filtered = df[
    (df['ConvertedCompYearly'] > threshold) &
    (df['RemoteWork'] == 'Remote')
]

# убираем пропуски в индустрии
filtered = filtered.dropna(subset=['Industry'])

# считаем популярные индустрии
top_industries = filtered['Industry'].value_counts()

top_industries

,count
Industry,
Software Development,1186
Fintech,190
Healthcare,188
Other:,176
"Internet, Telecomm or Information Services",138
Banking/Financial Services,88
Government,78
Media & Advertising Services,75
Retail and Consumer Services,65


**Вывод:** Среди высокооплачиваемых удалённых специалистов наиболее популярной индустрией является Software Development. Также значительную долю занимают Fintech и Healthcare. Это показывает, что IT-сфера остаётся лидером по уровню доходов среди remote специалистов.